경북대 BE 김진현 - Step 1: 유효성 검사 및 예외 처리

kimjhyun0627 opened the pull request:
## 구현한 기능

- 상품 이름 길이 유효성 검사
- 상품 이름 특수문자 유효성 검사
- 상품 이름 특수 패턴('카카오')에 대한 예외 처리
- 유효성 검사 관련 테스트 로직 작성(ProductControllerTest.java)
- 유효성 검사 실패 시 응답을 위한 Error Handler 구현(GlobalExeptionHandler.java)
- Error Handler 관련 테스트 로직 작성

Q: 에러 핸들러의 위치를 일단은 exception 패키지 내부에 잡았는데, 기존에 존재하던 Exception들과의 관계가 모호해진 것 같습니다. 이런 경우 상위 패키지에서 다른 패키지로 분리하여 핸들러를 위치시켜야 하는지, 그렇다면 그 패키지명은 통상적으로 어떻게 설정하는지 궁금합니다!

kimjhyun0627 committed:
docs: 기능 요구 사항 업데이트

- 기능 요구 사항 README에 업데이트

kimjhyun0627 committed:
feat: 상품 이름 길이 유효성 검사 구현

- 상품 이름 길이 유효성 검사 구현
- 유효성 검사 실패 시 예외 핸들러 구현
- 유효성 검사 관련 테스트 코드 생성
- README.md 업데이트

kimjhyun0627 committed:
feat: 상품 이름 특수문자 유효성 검사 구현

- 상품 이름 특수문자 유효성 검사 구현
- 특수문자 유효성 검사 관련 테스트 코드 생성
- README.md 업데이트

kimjhyun0627 committed:
feat: 상품 이름 특수 패턴 예외처리 구현

- 상품 이름 특수 패턴 예외 처리 구현: '카카오' 가 들어간 문자열 예외 처리
- 특수 패턴 유효성 검사 관련 테스트 코드 생성
- README.md 업데이트

kimjhyun0627 committed:
test: E2E 테스트 로직 추가

- 상품 이름 길이 유효성 검사 로직
- 상품 이름 특수문자 유효성 검사 로직
- 특수 문자열('카카오') 예외 처리 로직

kimjhyun0627 commented on src/main/java/gift/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+package gift.exception.handler;
```
Q: 에러 핸들러의 위치를 일단은 exception 패키지 내부에 잡았는데, 기존에 존재하던 Exception들과의 관계가 모호해진 것 같습니다. 이런 경우 상위 패키지에서 다른 패키지로 분리하여 핸들러를 위치시켜야 하는지, 그렇다면 그 패키지명은 통상적으로 어떻게 설정하는지 궁금합니다!

brainbackdoor replied:
- 패키지도 하나의 추상입니다. ExceptionHandler의 역할은 무엇인가요? 기존 Exception들의 역할은 무엇인가요? 이들을 묶어야할 이유가 있다면 같은 패키지에 위치해야 하는 것이고 다른 구성요소와 묶일 맥락이 있다면 적절히 추상화를 해야 하는 것이지요. 통상적이란 건 없습니다. 그저 각자 경험에 기반한 표현일 뿐이고, 제가 경험한 바에서 exception을 추상화하는 기준은 각 팀별로 모두 달랐어요. ExceptionHandler를 다른 Controller의 설정들과 같이 위치시키는 경우도 있고, Core 모듈이 분리되어 존재하는 경우도 있고, exception이 도메인 패키지 내에 위치하기도 하고, 공통 패키지로 분리되기도하고, 클린아키텍처 각 레이어 내에 존재하기도 하고 등등등등 그것도 한번 위치하면 변경되지 않고 늘 그곳에 있는게 아니라 맥락에 따라 리팩토링되는게 당연합니다. 우리가 풀고자 하는 문제는 머물러있지 않기 때문이죠. 만약 그런거라면, 스켈레톤 뼈대 코드 다 작성해두고 나머지 변경될 요소만 작성하면 될텐데, 많은 사람들이 공통적으로 쓰는 규약이 없다면 그 나름의 이유가 있는거라고 생각하시면 되어요.

brainbackdoor commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -1,14 +1,23 @@
-        @NotBlank(message = "상품 이름은 필수 입력값입니다.") String name,
-        @Min(value = 1, message = "가격은 1원 이상이어야 합니다.") int price,
-        @NotBlank(message = "이미지 URL은 필수 입력값입니다.") String imageUrl
+        @NotBlank(message = "상품 이름은 필수 입력값입니다.")
+        @Size(max=15, message = "상품 이름은 공백 포함 최대 15자까지 입력할 수 있습니다.")
+        @Pattern(regexp = "^[\\p{IsHangul}\\p{IsHan}\\p{IsAlphabetic}\\d\\s()\\[\\]+\\-&/_]+$",
+                message = "상품 이름에 허용되지 않는 문자가 포함되었습니다.")
+        @Pattern(regexp = "^(?!.*카카오).*",
+                message = "‘카카오’ 문구는 담당 MD와 협의 후에만 사용할 수 있습니다.")
+        String name,
+        @Min(value = 1, message = "가격은 1원 이상이어야 합니다.")
+        int price,
+        @NotBlank(message = "이미지 URL은 필수 입력값입니다.")
+        String imageUrl
```
- Controller에서 입력값 유효성 검증을 하였더라도, 불변식으로 보장되어야 하는 경우엔 생성자에서 유효성 검증은 반드시 진행하는게 좋다고 생각해요 (Controller 에서의 검증은 옵셔널하다고 생각해요) 정책이 이원화되어 있으면, 시스템이 상품을 등록하거나, 그 외 다른 맥락으로 상품 생성로직이 생길 때는 정책을 논할 때와 시점이 많이 달라져서 놓칠 수 있거든요.
- 기능 요구사항에 대해서는 해석의 여지가 있는 거 같습니다. 
  - 이름에 "카카오" 문구가 있는 경우 예외 처리를 한다. (오프라인에서 MD와 협의한 후 DB에 직접 등록)
  - 관리자 화면으로는 가능하고 API로는 불가능하다.
  - 동륵은 허용하되 상품 노출은 되지 않는다. (상태를 가지고 있고, MD 협의 후 상태를 노출 상태로 업데이트) 
  - ...
- 뭐가 되었든간에, 객체지향설계에선 불변식이 객체에 위치해야 합니다. 그렇지 않으면, 응집도가 낮아지는데요. 가령, 상품 등록 로직이 API, Consumer, Scheduler 등 요청자가 달라지거나, 시스템 호출 등 맥락이 달라지면 그에 따라 해당 정책이 산재하게 됩니다. 그럼 코딩 자체는 사람이 하는 일이다 보니, 실수가 생기기 마련입니다. 그래서 인스턴스화된 객체는 정책을 준수하고 있다고 보장해주도록 생성자에서 검증되는 것이 여러모로 좋습니다.

kimjhyun0627 replied:
리뷰요청 이후에 이 부분에 대해 생각해 보았는데, 코멘트로라도 남길 걸 그랬습니다...수정 반영하겠습니다 !
- 추가로 테이블 스키마에서의 name 필드 길이 역시 수정해 한 번 더 확인할 수 있도록 했습니다 !
- 기능 요구사항에 대해서는 깊이 생각해보지 못했는데, 말씀해주신 방안들 중 두 번째가 적합하다고 개인적으로 생각해 코드를 수정했습니다!
  → 그 과정에서 코드 중복을 최소화하기 위해 validationGroup을 분리하고, Consumer와 BiFunction 인터페이스를 사용해보았습니다..! 둘 다 처음 써 본 것들이라 아직 미흡한 점이 많을 것 같습니다...편하게 피드백해주시면 최대한 열심히 흡수해보겠습니다 !

kimjhyun0627 replied:
오버라이드한 mapRow()에서의 팩토리 메소드 사용에 대해서도 고민을 해 보았는데, 이미 DB에 저장된 것이라면 합당한 루트를 통해 저장되었을 것이고, 이를 다시 DB에서 꺼내와 매핑하는 것이라고 생각해 createPermittedProduct() 메소드를 사용했습니다 !

kimjhyun0627 replied:
다시 생각해보니, Consumer나 Bifunction을 사용함으로써 얻는 이점도 물론 있지만,
- 서비스 레이어와 컨트롤러 레이어의 중복 검증
- 코드 복잡성의 증가
이로 인해 잃는 것이 더 크다고 생각했고, 아직 스프링이 어색한 제게 조금 더 명시적인 방법을 사용하는 것이 맞다고 생각해, 뒤늦은 커밋으로 ValidationMode Enum을 생성해 서비스 로직에서 엔티티를 생성하며 검증을 진행하는 방식으로 변경했습니다 !
오래 생각해본 후 더 직관적이고 간결한 코드라고 생각해 변경하였으나, 어떤 로직이 더 나은지 아직은 잘 모르겠어서 이 부분 조심스럽게 코멘트 부탁드려봅니다...!

brainbackdoor replied:
무슨 말씀을 하시는지 잘 이해가 안되네요. 너무 문제를 어렵게 접근하고 계신거 같아요.

brainbackdoor commented on src/main/java/gift/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.HashMap;
+import java.util.Map;
+
+@RestControllerAdvice
+public class GlobalExceptionHandler {
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public Map<String, String> handleValidationExceptions(
+            MethodArgumentNotValidException ex
+    ) {
+        Map<String, String> errors = new HashMap<>();
+        for (FieldError err : ex.getBindingResult().getFieldErrors()) {
+            errors.put(err.getField(), err.getDefaultMessage());
+        }
+        return errors;
+    }
+}
```
- ExceptionHandler를 작성하셨다면, Custom Exception 에 대한 핸들링 외에도 상위의 Exception이나 공통적으로 응답되어야 할 HTTP Status에 대한 대응들도 작성해두시는 건 어떨까요?
- 이 Exception은 어디서 사용되나요? 현재 PR에는 관련 코드가 없는데요. 혹시 커밋이 모두 포함되지 않은걸까요?

kimjhyun0627 replied:
bean validation에 실패했을 때 MethodArgumentNotValidException이 발생한다고 알고 있어, 해당 exception만 우선적으로 처리했습니다..!
ResourceNotFoundException의 경우에는 분명 넣었다고 생각했는데...어딘가로 누락된 것 같습니다. ResourceNotFoundException을 포함한 프로젝트 전체의 Exception에 대한 핸들러 구현하겠습니다 !

handler 패키징에 대해서도 생각해본 결과, 아직 프로젝트의 크기가 크지 않기도 하거니와, 모든 exception을 한 클래스 안에서 처리하는 지금의 형태로는 exception 패키지 내부에 위치시키는게 가장 직관적일 것이라고 생각해 해당 위치로 리팩토링 하였습니다. 추후 custom exception이 더 많이 생기더라도, exception 패키지 내부에서 custom exception을 패키징하는 형식을 가지는 것이 지금의 생각입니다.

brainbackdoor replied:
- 아, 스프링이 오랜만이라 Bean Validation 예외 이름이 헷갈렸군요. 그렇다면, 예외가 발생할 수 있는 케이스가 다양하니 응답을 좀 더 구조화하는게 좋겠어요

```java
public record ErrorResponse(
    String code,
    String message,
    Map<String, List<String>> fieldErrors,
    LocalDateTime timestamp
) {
    public static ErrorResponseBuilder builder() {
        return new ErrorResponseBuilder();
    }
}

@ExceptionHandler(MethodArgumentNotValidException.class)
@ResponseStatus(HttpStatus.BAD_REQUEST)
public ErrorResponse handleValidationExceptions(MethodArgumentNotValidException ex) {
    List<FieldErrorDetail> fieldErrors = ex.getBindingResult().getFieldErrors()
            .stream()
            .map(error -> new FieldErrorDetail(
                error.getField(),
                error.getDefaultMessage(),
                error.getRejectedValue(),
                error.getCode()
            ))
            .collect(Collectors.toList());
    
    return ErrorResponse.builder()
            .code("VALIDATION_FAILED")
            .message("입력값 검증에 실패했습니다.")
            .fieldErrors(fieldErrors)
            .timestamp(LocalDateTime.now())
            .build();
}

public record FieldErrorDetail(
    String field,
    String message,
    Object rejectedValue,
    String errorCode
) {}
```

brainbackdoor commented on src/test/java/gift/controller/AdminProductControllerTest.java:
```diff
@@ -78,6 +78,31 @@ void createProduct_Success_and_MissingField_Fail() {
         HttpEntity<MultiValueMap<String, String>> badEntity = new HttpEntity<>(badForm, headers);
         ResponseEntity<String> fail = restTemplate.postForEntity(baseUrl, badEntity, String.class);
         assertThat(fail.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
+
+        MultiValueMap<String, String> badForm2 = new LinkedMultiValueMap<>();
+        badForm2.add("name", "invalid!!");
+        badForm2.add("price", "1");
+        badForm2.add("imageUrl", "http://img");
+        HttpEntity<MultiValueMap<String, String>> badEntity2 = new HttpEntity<>(badForm2, headers);
+        ResponseEntity<String> fail2 = restTemplate.postForEntity(baseUrl, badEntity2, String.class);
+        assertThat(fail2.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
+
+        MultiValueMap<String, String> badForm3 = new LinkedMultiValueMap<>();
+        badForm3.add("name", "toolongtomakeittoproductname");
+        badForm3.add("price", "1");
+        badForm3.add("imageUrl", "http://img");
+        HttpEntity<MultiValueMap<String, String>> badEntity3 = new HttpEntity<>(badForm3, headers);
+        ResponseEntity<String> fail3 = restTemplate.postForEntity(baseUrl, badEntity3, String.class);
+        assertThat(fail3.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
+
+        MultiValueMap<String, String> badForm4 = new LinkedMultiValueMap<>();
+        badForm4.add("name", "카카오테크캠퍼스");
+        badForm4.add("price", "1");
+        badForm4.add("imageUrl", "http://img");
+        HttpEntity<MultiValueMap<String, String>> badEntity4 = new HttpEntity<>(badForm4, headers);
+        ResponseEntity<String> fail4 = restTemplate.postForEntity(baseUrl, badEntity4, String.class);
+        assertThat(fail4.getStatusCode()).isEqualTo(HttpStatus.BAD_REQUEST);
+
```
- 하나의 테스트에 여러 의도를 담아주지 마세요. 각 의도에 맞춰 테스트 케이스를 나누고 assertion 해주세요. 가령, 위 테스트는 각 케이스가 모두 실패하면 하나 수정하고 다시 돌려보고 하나 수정하고 다시 돌려보며 badForm4 까지 반복해나가야 합니다. 왜냐하면 앞에서 실패하면 뒤에는 왜 실패하는지 바로 인지가 안되기 때문이죠.

kimjhyun0627 replied:
수정하겠습니다 !

brainbackdoor reviewed (changes requested):
PR에 모든 작업이 다 포함되지 않은 거 같아서 Request Changes 합니다. 한번 확인해보셔요~

kimjhyun0627 committed:
refactor: 코드 리뷰 기반 수정

- 생성자 유효성 검증 로직 추가
- 기능 요구사항 변경: 관리자 화면으로는 '카카오'가 포함된 상품명 등록 가능, api로는 불가능
- 데이터 스키마 수정(name 필드 길이 15로 제한)
- 예외처리 핸들러 패키지 리팩토링 및 핸들러 추가
- 테스트 코드 세분화 및 구체화

kimjhyun0627 commented:
말씀해주신 부분들 바탕으로 수정 진행했습니다..! 감사합니다 !
- 객체 생성 시 팩토리 매서드 사용을 강제하는 방식으로 구현했는데, 그 과정에서 각 레이어가 조금은 지저분해진게 아닌가 생각됩니다.

kimjhyun0627 committed:
refactor: controller와 service 역할 분리

- 생성자 유효성 검증 로직 변경
- 기능 요구사항 변경: 관리자 화면으로는 '카카오'가 포함된 상품명 등록 가능, api로는 불가능

kimjhyun0627 committed:
refactor: service: updateProduct minor change

- 기존: id에 해당하는 Product가 있는지만 확인 후, 새 Product를 만들어 덮어쓰기
- 변경: id에 해당하는 Product를 가져와, withField() 메소드 호출을 통해 값 수정

brainbackdoor commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,33 +1,106 @@
+
+    private static final int MAX_NAME_LENGTH = 15;
+
+    public static Product createProduct(
+            Long id,
+            String name,
+            int price,
+            String imageUrl,
+            ValidationMode mode
+    ) {
+        if (id != null) {
+            validateId(id);
+        }
+        validateName(name, mode);
+        validatePrice(price);
+        validateImageUrl(imageUrl);
+        return new Product(id, name, price, imageUrl);
+    }
```
- 원시값을 포장하는건 어떨까요
```
public record Product(
    ProductName name,      
    ProductPrice price,   
    ImageUrl imageUrl     
...
) {}
```

- Entity와 Value Object의 차이에 대해 학습해보시겠어요? 그 맥락에서 Entity는 Record와 어떤 부분이 어울리지 않는지 한번 고민해보시겠어요ㅗ?

kimjhyun0627 replied:
가변성이 필요하고, ORM의 제약이 생깁니다. 또한 영속성 관리 측면에서 엔티티와 어울리지 않는 것 같습니다.
지난 챕터 멘토님께서 레코드 사용을 추천해주시기도 하였고, 저 역시 레코드 사용이 처음이라 연습 겸 이번 코드리뷰 바탕으로 value object 역시 레코드로 작성했습니다. 혹시 이대로 진행해도 괜찮을까요?

kimjhyun0627 replied:
update: value object는 레코드로, domain은 entity로 변경했습니다 !

brainbackdoor commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,33 +1,106 @@
+                    "상품명에 포함된 특정 문구는 담당 MD와 협의 후에만 사용할 수 있습니다."
+            );
+        }
+    }
```
- 위에서 제안한대로 메서드로 분리하기 보다 Value Object를 도출하고 생성자에서 이를 검증하는 것은 어떨까요? 지금과 같이 검증로직이 계속 늘어나면 점점 유지보수가 어려워집니다.

brainbackdoor commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,33 +1,106 @@
+            throw new IllegalArgumentException("이미지 URL은 필수 입력값입니다.");
+        }
+    }
+
+    private static void validatePattern(String name) {
+        boolean valid = MANAGEMENT_PATTERNS.stream()
+                .allMatch(pattern -> pattern.matcher(name).matches());
+        if (!valid) {
+            throw new IllegalArgumentException(
+                    "상품명에 포함된 특정 문구는 담당 MD와 협의 후에만 사용할 수 있습니다."
+            );
+        }
+    }
+    
     public Product withId(Long id) {
+        validateId(id);
         return new Product(id, this.name, this.price, this.imageUrl);
     }
 
-    public Product withName(String name) {
+    public Product withName(String name, ValidationMode validationMode) {
+        validateName(name, validationMode);
         return new Product(this.id, name, this.price, this.imageUrl);
     }
 
     public Product withPrice(int price) {
+        validatePrice(price);
         return new Product(this.id, this.name, price, this.imageUrl);
     }
 
     public Product withImageUrl(String imageUrl) {
+        validateImageUrl(imageUrl);
         return new Product(this.id, this.name, this.price, imageUrl);
     }
+
+    public enum ValidationMode {
+        NORMAL,
+        PERMITTED,
+        DATABASE
+    }
```
- 음.. Presentation Layer와 도메인 모델 결합도가 너무 증가했는데요. 유효성에 대한 기준을 객체에서 판단하지 않는 것도 그렇구요. 
- admin에서는 가능하고, 일반 사용자들이 사용하는 api에서는 불가능해야 한다고 Comment 남겼던 것 때문에 이렇게 작성하신거라면, 의도가 잘 전달되지 않은 거 같네요. 권한은 인가정책에 해당합니다.

kimjhyun0627 replied:
특수 패턴 로직을 
기존: rest api에서 '카카오' 문자 입력 불가
변경: rest api에서 '카카오' 문자 입력 시 숨김 처리
admin 페이지에서 이를 숨김 해제할 수 있음
으로 변경함으로써 조금은 결합도를 감소시키려고 노력했습니다..! 제 머리로는 도저히 기본 로직을 유지하면서 결합도를 감소시킬 방법이 떠오르지 않았습니다...
그럼에도, 기존 정보를 다시 불러오는 과정에서, service가 어느 controller에서 호출된 것인지 알 수 없기에, 숨김 처리 로직이 아직 controller에 있습니다...ㅜ

brainbackdoor commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -1,14 +1,30 @@
-        @Min(value = 1, message = "가격은 1원 이상이어야 합니다.") int price,
-        @NotBlank(message = "이미지 URL은 필수 입력값입니다.") String imageUrl
-) {
+        @NotBlank(message = "상품 이름은 필수 입력값입니다.",
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Size(max = 15, message = "상품 이름은 공백 포함 최대 15자까지 입력할 수 있습니다.",
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Pattern(regexp = "^[\\p{IsHangul}\\p{IsHan}\\p{IsAlphabetic}\\d\\s()\\[\\]+\\-&/_]+$",
+                message = "상품 이름에 허용되지 않는 문자가 포함되었습니다.",
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Pattern(regexp = "^(?!.*카카오).*",
+                message = "‘카카오’ 문구는 담당 MD와 협의 후에만 사용할 수 있습니다.",
+                groups = ValidationGroups.DefaultGroup.class)
+        String name,
```
- DefaultGroup의 역할은 무엇인가요? PermittedGroup의 역할은 무엇인가요? 기존과 달리 현재 구조가 갖는 이점은 무엇인가요?

kimjhyun0627 replied:
controller 단계에서 특수 패턴 검증 로직을 제거함으로써 자연스럽게 group도 제거했습니다 !

brainbackdoor commented on src/main/java/gift/repository/ProductRowMapper.java:
```diff
@@ -16,7 +16,8 @@ public Product mapRow(ResultSet rs, int rowNum) throws SQLException {
                 rs.getInt("price"),
-                rs.getString("image_url")
+                rs.getString("image_url"),
+                ValidationMode.DATABASE
```
- NORMAL 외에는 동작하는 경우가 없는거 같은데요. 로깅 목적도 아니고, 별도의 타입을 둔 이유가 무엇인가요?

kimjhyun0627 replied:
원래 의도는 "service 레이어까지 인가 레벨을 끌고 내려가자" 였는데, 앞서 사용한 group과 완전히 동일한 로직이라 동작하지 않았던 것 같습니다.

brainbackdoor commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -68,4 +70,8 @@ public String delete(@PathVariable("id") Long id) {
         return "redirect:/admin/products";
     }
+
+    private Product toEntity(ProductRequest r) {
+        return new Product(null, r.name(), r.price(), r.imageUrl());
+    }
```
- 변환 로직은 dto에 위치해주세요. controller의 역할도 아니니까요. 그나저나 정적 팩토리 메서드를 작성했다면 적극적으로 활용하는 것은 어떨까요? 이 부분은 ValidationMode 가 누락되었네요. 이처럼 정책이 분산되면 휴먼 에러가 발생할 여지가 있어요.

brainbackdoor commented on src/main/java/gift/dto/ProductRequest.java:
```diff
@@ -1,14 +1,30 @@
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Size(max = 15, message = "상품 이름은 공백 포함 최대 15자까지 입력할 수 있습니다.",
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Pattern(regexp = "^[\\p{IsHangul}\\p{IsHan}\\p{IsAlphabetic}\\d\\s()\\[\\]+\\-&/_]+$",
+                message = "상품 이름에 허용되지 않는 문자가 포함되었습니다.",
+                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
+        @Pattern(regexp = "^(?!.*카카오).*",
```
```suggestion
        @Pattern(regexp = ProductNamePatterns.ALLOWED_CHARACTERS,
                message = "상품 이름에 허용되지 않는 문자가 포함되었습니다.",
                groups = {ValidationGroups.DefaultGroup.class, ValidationGroups.PermittedGroup.class})
        @Pattern(regexp =ProductNamePatterns.KAKAO_RESTRICTION,
```
- 정규식 패턴을 다른 곳에서도 활용하고 있던데요. 정책이 분산되어 있으면 나중에 실수 할 수 있어요. 그리고 이런 부분은 static final로 선언해서 재사용하는건 어떨까요?

brainbackdoor commented on src/test/java/gift/controller/AdminProductControllerTest.java:
```diff
@@ -36,98 +37,279 @@ public class AdminProductControllerTest {
-        assertThat(response.getBody()).contains("신규 상품 등록");
-        assertThat(response.getBody()).contains("name=\"name\"");
-    }
+        @Test
+        @DisplayName("GET list page - 상품 목록 조회")
+        void listPage_ShouldReturnProductsInModel() {
+            productRepository.save(new Product(null, "A", 10, "uA"));
+
+            ResponseEntity<String> response = restTemplate.getForEntity(baseUrl, String.class);
+
+            assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+            assertThat(response.getBody()).contains("상품 목록");
+            assertThat(response.getBody()).contains("A");
+        }
```
- Controller Test지만, 실제로는 값 검증을 하고 있군요. 이렇게 여러 Component가 협력하는 테스트는 통합테스트라고 표현하는 편입니다. 그렇다하더라도 repository를 주입해서 저장하기 보다는, Test Fixture를 따로 둬서 재사용합니다. (URL을 테스트 내에서 하드코딩해서 요청하고 있던데요. 우선 이 테스트에서 RestTemplate 의존을 제거해보는 건 어떨까요?)

brainbackdoor commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,33 +1,106 @@
+
+    private static final int MAX_NAME_LENGTH = 15;
+
+    public static Product createProduct(
+            Long id,
+            String name,
+            int price,
+            String imageUrl,
+            ValidationMode mode
+    ) {
+        if (id != null) {
+            validateId(id);
+        }
+        validateName(name, mode);
+        validatePrice(price);
+        validateImageUrl(imageUrl);
+        return new Product(id, name, price, imageUrl);
+    }
```
- 정적 팩토리 메서드는 하나여야만 할까요? id가 없이 생성되는 경우가 있다면 이를 위한 팩토리 메서드도 대응해주면 되지않을까요?

brainbackdoor reviewed (changes requested):
테스트 코드 상세히 나눠주셨군요! 👍🏻  하지만, 도메인 모델에 대한 테스트가 존재하지 않아서 많이 아쉽네요~ 
지난 Comment 에서 몇가지 오해가 있었는지, 제가 의도했던 것과 방향이 많이 달라진듯하여 몇몇 Comment를 남겨두었어요. 혹시 설명이 좀 더 필요하다면 DM 주세요~
좋은 밤 보내시구요~

kimjhyun0627 committed:
refactor: 2차 코드리뷰 바탕 수정

- 특수 패턴(카카오) 로직 변경
- 기존: rest api에서 '카카오' 문자 입력 불가
- 변경: rest api에서 '카카오' 문자 입력 시 숨김 처리, admin 페이지에서 이를 숨김 해제할 수 있음

- Value Object 도출(record 형식)
- 불필요한 로직 삭제
- 테스트 코드 변경(E2E 명시)

kimjhyun0627 committed:
refactor: 2차 코드리뷰 바탕 수정

- 특수 패턴(카카오) 로직 변경
- 기존: rest api에서 '카카오' 문자 입력 불가
- 변경: rest api에서 '카카오' 문자 입력 시 숨김 처리, admin 페이지에서 이를 숨김 해제할 수 있음

- Value Object 도출(record 형식)
- 불필요한 로직 삭제
- 테스트 코드 변경(E2E 명시)

kimjhyun0627 committed:
feat: 도메인 테스트 추가

- 도메인 테스트 추가

kimjhyun0627 committed:
refactor: Product domain entity로 변경

- 기존: record
- 변경: entity

kimjhyun0627 committed:
refactor: value object 시그니처 변경

- 기존: value
- 변경: id, name, price, url

kimjhyun0627 committed:
feat: service 테스트 코드 추가

- service 레이어 테스트 코드 추가

kimjhyun0627 committed:
refactor: controller test / e2e test 분리

- controller test / e2e test 분리

kimjhyun0627 committed:
feat: repository 테스트 구현

- repository test 구현

kimjhyun0627 committed:
fix: AdminProductControllerTest: mockbean 사용 수정

- AdminProductControllerTest: mockbean -> mockitobean

brainbackdoor commented on src/main/java/gift/controller/AdminProductController.java:
```diff
@@ -0,0 +1,117 @@
+            BindingResult bindingResult,
+            HttpServletResponse response
+    ) {
+        if (bindingResult.hasErrors()) {
+            response.setStatus(HttpStatus.BAD_REQUEST.value());
+            return "admin/product_form";
+        }
```
- 중복 코드가 발생하는군요. 리팩토링하려면어떻게 하는게 좋을까요?

brainbackdoor commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -32,46 +33,48 @@ public ProductController(ProductService svc) {
-        return new ProductResponse(
-                e.id(), e.name(), e.price(), e.imageUrl()
-        );
+    private Product findUnHiddenProduct(Long id) {
+        Product p = productService.getProductById(id)
+                .orElseThrow(() -> new ProductNotFoundExection(id));
+        if (p.hidden()) {
+            throw new ProductHiddenException(id);
+        }
```
- hidden인데, 예외로 알려주면 안되지 않을까요. brute force로 어떤 id들이 숨겨져 있는지 탐색할 여지가 있어보여서요.

brainbackdoor commented on src/main/java/gift/entity/product/Product.java:
```diff
@@ -0,0 +1,116 @@
+        this.hidden = hidden;
+    }
+
+    public static Product of(Long id, String name, int price, String imageUrl, boolean hidden) {
+        return new Product(
+                new ProductId(id),
+                new ProductName(name),
+                new ProductPrice(price),
+                new ProductImageUrl(imageUrl),
+                hidden
+        );
+    }
+
+    public static Product of(Long id, String name, int price, String imageUrl) {
+        return new Product(
+                new ProductId(id),
+                new ProductName(name),
+                new ProductPrice(price),
+                new ProductImageUrl(imageUrl),
+                false
+        );
+    }
+
+    public static Product of(String name, int price, String imageUrl) {
+        return new Product(
+                null,
+                new ProductName(name),
+                new ProductPrice(price),
+                new ProductImageUrl(imageUrl),
+                false
+        );
+    }
```
```suggestion
    public static Product of(Long id, String name, int price, String imageUrl, boolean hidden) {
        return new Product(
                new ProductId(id),
                new ProductName(name),
                new ProductPrice(price),
                new ProductImageUrl(imageUrl),
                hidden
        );
    }

    public static Product of(Long id, String name, int price, String imageUrl) {
        return of(id, name, price, imageUrl, false);
    }

    public static Product of(String name, int price, String imageUrl) {
        return of(null, name, price, imageUrl, false);
    }
```
- 중복코드는 습관적으로 제거해주세요.

brainbackdoor commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -1,58 +1,89 @@
 @Transactional
 public class ProductServiceImpl implements ProductService {
 
+    private static final List<Pattern> FORBIDDEN_PATTERNS = List.of(
+            Pattern.compile("카카오")
+    );
```
- 이 부분은 도메인 모델로 옮겨져야 해요. Service Layer에서 비즈니스 규칙 두는걸 지양해주세요.

brainbackdoor commented on src/main/java/gift/entity/product/value/ProductImageUrl.java:
```diff
@@ -0,0 +1,35 @@
+            Pattern.CASE_INSENSITIVE
+    );
+
+    public ProductImageUrl {
+        Objects.requireNonNull(url, "이미지 URL은 필수 입력값입니다.");
+        String trimmed = url.trim();
+        if (trimmed.isEmpty()) {
+            throw new IllegalArgumentException("이미지 URL은 필수 입력값입니다.");
+        }
+        if (!URL_PATTERN.matcher(trimmed).matches()) {
+            throw new IllegalArgumentException("올바른 HTTP/HTTPS 이미지 URL이어야 합니다.");
+        }
+        try {
+            URI uri = new URI(trimmed);
+            String scheme = uri.getScheme().toLowerCase();
+            if (!"http".equals(scheme) && !"https".equals(scheme)) {
+                throw new IllegalArgumentException("이미지 URL은 HTTP 또는 HTTPS 프로토콜이어야 합니다.");
+            }
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("올바른 이미지 URL이어야 합니다.", e);
+        }
+        url = trimmed;
+    }
```
```suggestion
    private static final String NULL_URL_MESSAGE = "이미지 URL은 필수 입력값입니다.";
    private static final String INVALID_URL_MESSAGE = "올바른 HTTP/HTTPS 이미지 URL이어야 합니다.";
    private static final String INVALID_PROTOCOL_MESSAGE = "이미지 URL은 HTTP 또는 HTTPS 프로토콜이어야 합니다.";

    public ProductImageUrl {
        Objects.requireNonNull(url, NULL_URL_MESSAGE);
        String trimmed = url.trim();
        
        validateUrlFormat(trimmed);
        validateUrlProtocol(trimmed);
        
        url = trimmed;
    }
    
    private static void validateUrlFormat(String url) {
        if (url.isEmpty()) {
            throw new IllegalArgumentException(NULL_URL_MESSAGE);
        }
        
        if (!URL_PATTERN.matcher(url).matches()) {
            throw new IllegalArgumentException(INVALID_URL_MESSAGE);
        }
    }
    
    private static void validateUrlProtocol(String url) {
        try {
            URI uri = new URI(url);
            String scheme = uri.getScheme();
            
            if (scheme == null || (!"http".equalsIgnoreCase(scheme) && !"https".equalsIgnoreCase(scheme))) {
                throw new IllegalArgumentException(INVALID_PROTOCOL_MESSAGE);
            }
        } catch (URISyntaxException e) {
            throw new IllegalArgumentException(INVALID_URL_MESSAGE + " - " + e.getMessage(), e);
        }
    }
```
```java
@Test
@DisplayName("유효한 HTTPS URL이면 정상 생성된다")
void shouldCreateProductImageUrl_withValidHttpsUrl() {
    // given
    String validHttpsUrl = "https://www.example.com/path/to/resource.jpg";
    
    // when
    ProductImageUrl productImageUrl = new ProductImageUrl(validHttpsUrl);
    
    // then
    assertThat(productImageUrl.url()).isEqualTo(validHttpsUrl);
}

@Test
@DisplayName("null URL이면 예외가 발생한다")
void shouldThrowException_whenUrlIsNull() {
    assertThatThrownBy(() -> new ProductImageUrl(null))
            .isInstanceOf(NullPointerException.class)
            .hasMessage("이미지 URL은 필수 입력값입니다.");
}

@Test
@DisplayName("빈 문자열이면 예외가 발생한다")
void shouldThrowException_whenUrlIsEmpty() {
    assertThatThrownBy(() -> new ProductImageUrl("   "))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("이미지 URL은 필수 입력값입니다.");
}

@Test
@DisplayName("잘못된 URL 형식이면 예외가 발생한다")
void shouldThrowException_whenUrlFormatIsInvalid() {
    assertThatThrownBy(() -> new ProductImageUrl("invalid-url"))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("올바른 HTTP/HTTPS 이미지 URL이어야 합니다.");
}

@Test
@DisplayName("FTP 프로토콜이면 예외가 발생한다")
void shouldThrowException_whenProtocolIsFtp() {
    assertThatThrownBy(() -> new ProductImageUrl("ftp://example.com/image.jpg"))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessage("이미지 URL은 HTTP 또는 HTTPS 프로토콜이어야 합니다.");
}

@Test
@DisplayName("URI 구문 오류가 있으면 예외가 발생한다")
void shouldThrowException_whenUriSyntaxIsInvalid() {
    assertThatThrownBy(() -> new ProductImageUrl("http://exa mple.com"))
            .isInstanceOf(IllegalArgumentException.class)
            .hasMessageContaining("올바른 HTTP/HTTPS 이미지 URL이어야 합니다.");
}
```
- 객체를 어떻게 나눠야 할지 모르겠을 때는 우선 비슷해보이는 기능끼리 묶어서 메서드 추출을 해보세요. 이 때, input과 output 을 예상하고 테스트 코드로 보호해보세요. 가령 위 로직은 예외 케이스에 대해서 메서드로 분리하고 테스트를 작성해가면서 리팩토링해갈 수 있어요. 현재는 변경 지점이 너무 많아 보입니다.

brainbackdoor commented on src/main/java/gift/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,91 @@
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleIllegalArgument(IllegalArgumentException ex) {
+        return new ErrorResponse(
+                "BAD_REQUEST",
```
- 값으로 관리하기 보단 ENUM을 적극적으로 활용해보세요.

brainbackdoor commented on src/main/java/gift/service/ProductServiceImpl.java:
```diff
@@ -1,58 +1,89 @@
+                .orElseThrow(() -> new ProductNotFoundExection(id));
+
+        Product updatedProduct = existing.withName(name)
+                .withPrice(price)
+                .withImageUrl(imageUrl);
+        if (isForbidden(name)) {
+            updatedProduct = updatedProduct.withHidden(true);
+        }
 
-        Product updated = existing.withName(product.name())
-                .withPrice(product.price())
-                .withImageUrl(product.imageUrl());
-        return repo.save(updated);
+        return repo.save(updatedProduct);
     }
 
     @Override
     public void deleteProduct(Long id) {
         if (!repo.existsById(id)) {
-            throw new ResourceNotFoundException("상품을 찾을 수 없습니다: " + id);
+            throw new ProductNotFoundExection(id);
         }
         repo.deleteById(id);
     }
+
+    @Override
+    public void hideProduct(Long id) {
+        Product p = repo.findById(id)
+                .orElseThrow(() -> new ProductNotFoundExection(id));
+        p = p.withHidden(true);
+        repo.save(p);
+    }
+
+    @Override
+    public void unhideProduct(Long id) {
+        Product p = repo.findById(id)
+                .orElseThrow(() -> new ProductNotFoundExection(id));
+        p = p.withHidden(false);
+        repo.save(p);
+    }
```
- 일단 숨김처리하고, unhide 하는 프로세스인가 보군요. 아직 인증/인가를 어떻게 구현할지 학습 못하셨나봐요. 다음에 학습하게 되면 다시 과제를 수행해봐도 좋겠어요~

brainbackdoor commented on src/main/resources/db/init/schema.sql:
```diff
@@ -1,7 +1,8 @@
     PRICE     INT          NOT NULL,
-    IMAGE_URL VARCHAR(255) NOT NULL
+    IMAGE_URL VARCHAR(255) NOT NULL,
+    HIDDEN    BOOLEAN      NOT NULL DEFAULT FALSE
```
- 이럴 경우엔 보수적으로 숨김처리를 true로 두는 편입니다.

brainbackdoor commented on src/test/java/gift/e2e/ProductE2ETest.java:
```diff
@@ -0,0 +1,207 @@
+
+        @Test
+        void listExcludesHidden() throws Exception {
+            ProductFixture.save(productRepository, "V1", 10, "http://example.com/v1.png");
```
- Fixture에 인자로 repository를 주입하는 것은 어색하네요. 공통적으로 사용하는 로직이 있다면 이를 상속하는 것은 어떨까요? 상속이 싫다면, SpringBootTest의 경우 Spring Bean 주입시 필요한 빈들 주입하면서 사전로직을 수행시키는 여러 방법들이 있으니 이를 학습해보셔도 좋을 거 같네요.

brainbackdoor reviewed (approved):
피드백 반영하느라 고생하셨어요~ 👍🏻 
몇가지 간단한 Comment 남겨두었으니 참고바래요~
